### PR TITLE
Fix: ref type in Animated Component

### PR DESCRIPTION
**Pull request by nnnnnoel**
Please fill in this template.

- [ ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [ ] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/react-native/index.d.ts [hunk 1]
```diff
@@ -8489,19 +8489,28 @@ export namespace Animated {
...
+
+  export type AnimatedProps<T> = { [key in keyof T]: WithAnimatedValue<T[key]> };
+
+  export interface AnimatedComponent<
+    T extends React.ComponentType<ComponentProps<T>> | React.Component<ComponentProps<T>>
+  >
+    extends React.FC<
+      AnimatedProps<ComponentProps<T>> & {
+        ref?:
+          | React.RefObject<AnimatedComponent<T> | undefined>
```
Comment (TheSavior) [inline types/react-native/index.d.ts]: Does this also include the ref callback form? `ref={(ref) => ...}`

Comment (zusinShinpei) [reply types/react-native/index.d.ts]: No, That's my mistake. please show next commit.

File: types/react-native/index.d.ts [hunk 2]
```diff
@@ -8489,19 +8489,26 @@ export namespace Animated {
...
-    }
+    extends ThisType<
+      T extends object
+        ? { [K in keyof Pick<T, Exclude<keyof T, 'ref'>>]?: WithAnimatedValue<T[K]> }
+        : T extends (infer P)[]
+        ? WithAnimatedValue<P>[]
+        : T | Value | AnimatedInterpolation
+    > {}
+
+  export type AnimatedProps<T> = { [key in keyof T]: WithAnimatedValue<T[key]> };
```
Comment (alloy) [inline types/react-native/index.d.ts]: Nit: For consistency, please keep types camelized:

```suggestion
  export type AnimatedProps<T> = { [Key in keyof T]: WithAnimatedValue<T[Key]> };
```

**Review by alloy — CHANGES_REQUESTED** [review]
Can you add test coverage for the things this fixes?

File: types/react-native/index.d.ts (hunk 2 above)
Comment (zusinShinpei) [reply types/react-native/index.d.ts]: resolved

**Comment by alloy** [comment]
@zusinShinpei Wanna make sure you didn’t miss my question for test coverage for the case this fixes?

**Comment by zusinShinpei** [comment]
@alloy I'll make that in few days

**Comment by zusinShinpei** [comment]
@alloy I couldn't acceptance about result of ci that 3.6 below version of typescript doesn't work in infer

**Comment by alloy** [comment]
The CI error occurs in the test. Is it a specific thing to the test or do the new changes really require all user projects to bump their TS version?

If it’s just this test, then perhaps it would be better to comment it out for now?

**Comment by zusinShinpei** [comment]
I cannot agree about this test, infer's specific change is very important problem. don't you think about infer that 3.5 or below version of typescript is wrong?

**Comment by alloy** [comment]
Hey sorry, missed your comment. Unfortunately I’m unsure if I understand you correctly; you _do_ think this version bump is essential?

**Comment by zusinShinpei** [comment]
@alloy  I thought the typescript@3.5 has unknown problem at infer syntax.If it is for real,  the users have to bump version of typescript. I could not find the problem in this PR.

**Comment by alloy** [comment]
I will try to take a look at this locally myself 👍

**Comment by alloy** [comment]
Doing so now.

**Pull request closed without merging**